### ISSUE TITLE
Should the new SuiteRuntimeClient extend ZMQClient?

### ISSUE THREAD

**Issue by kinow** [cylc/cylc-flow#3014]
While reviewing #3004 , and trying things in a Jupyter Notebook, I noticed that `SuiteRuntimeClient` is not really a `SuiteRuntimeClient`. There's a `__new__` method, that returns not a `SuiteRuntimeClient` object, but instead ` ZMQClient`.

Is that a recommendation of 0mq or pyzmq? Or was it to reduce the dependency on OO?

Here's a sample code to demonstrate it:

```python
from cylc.network.client import SuiteRuntimeClient, ZMQClient
client = SuiteRuntimeClient('five')
print(type(client))
print(isinstance(client, SuiteRuntimeClient))
print(issubclass(SuiteRuntimeClient, ZMQClient))
```

![image](https://user-images.githubusercontent.com/304786/54574387-fb940180-4a54-11e9-8c5f-2d942fbf0bce.png)

I think we could run into problems later when adding typing (though that's static time only, no runtime enforcers) or if/when we need to use `isinstance` for example. But if that's a requirement for 0mq I think we could just document why and close this issue.

**Comment by hjoliver** [comment]
Ping @oliver-sanders (presumably).

**Comment by dwsutherland** [comment]
I believe `SuiteRuntimeClient` is an interface, designed to be inherited. See [PoC-GraphQL](https://github.com/cylc/cylc/issues/2900#issuecomment-469199387)

**Comment by matthewrmshin** [comment]
(Totally off topic, but I really like [Go's Interface System](https://en.wikipedia.org/wiki/Go_(programming_language)#Interface_system). No dodgy inheritance issue!)

**Comment by oliver-sanders** [comment]
Yes `SuiteRuntimeClient` is a factory!

This was intentional, it's where I started implementing the SSH tunnelling mode, you call `SuiteRuntimeClient` and depending on the arguments you provide you get a `ZMQClient` or an `SSHClient` both offering the same interface (by duck typing).

**Comment by kinow** [comment]
>(Totally off topic, but I really like Go's Interface System. No dodgy inheritance issue!)

That looks amazing! No idea how they implement that in their low level compiler/parser.

**Comment by kinow** [comment]
>Yes SuiteRuntimeClient is a factory!

Ah! That makes more sense.

>This was intentional, it's where I started implementing the SSH tunnelling mode, you call SuiteRuntimeClient and depending on the arguments you provide you get a ZMQClient or an SSHClient both offering the same interface (by duck typing).

But that breaks OO and the `isinstance` and `issubclass`. Woudln't be better to perhaps have a `SuiteRuntimeFactory` or `SuiteRuntimeClientFactory`, that did not necessarily use `__new__`? Or that used `__new__` but returned subclasses of a common type?

While I understand OO is not as enforced in Python as in Java (and I may be mistaken here in trying to keep the methods `isinstanceof` and `issubclassof` working), I think it is doable to implement a factory that returns a SSH or a ZMQ client, without breaking OO, which sounds less risky for us? e.g. [this](https://python-3-patterns-idioms-test.readthedocs.io/en/latest/Factory.html) 

The [O'Reilly Python in a Nutshell](https://www.oreilly.com/library/view/python-in-a/9781491913833/ch04.html#factory-function_idiom) shows some ways to use `__new__` with and without returning the same type. But they mention too

>In other words, `__new__` is not constrained to return a new instance of C, although normally it’s expected to do so.

As well as the [Python documentation](https://docs.python.org/3/reference/datamodel.html#object.__new__)

>The return value of `__new__()` should be the new object instance (usually an instance of cls). (...) `__new__()` is intended mainly to allow subclasses of immutable types (like int, str, or tuple) to customize instance creation. It is also commonly overridden in custom metaclasses in order to customize class creation.

**Comment by oliver-sanders** [comment]
I don't think it matters much (we aren't going to do any further subclassing) but feel free to change to something more OO or more functional or just more pretty!

**Comment by matthewrmshin** [comment]
Do we just have a naming convention issue here? E.g. when we have a CamelCase name, we expect it to be a class? If `SuiteRuntimeClient` is a factory method/function, should it have a lower_case name instead?

**Comment by oliver-sanders** [comment]
Composition might be nicer here...

**Comment by matthewrmshin** [comment]
> Composition might be nicer here...

Could you please elaborate? I am not sure which way are we going?

**Comment by kinow** [comment]
>Composition might be nicer here...

+1 if we are able to achieve a set of functions/classes/etc that play together and prevents this possible issue with `isinstanceof` etc.

Even in Java the OO design of classes needs to be thoroughly reviewed to avoid mistakes from the past, such as the `java.util.Properties`, https://stackoverflow.com/a/11344295:

>The java.util.Properties class is a good example of a bad use of inheritance. Rather than using a Hashtable to store its properties, it extends Hashtable, in order to reuse its methods and to avoid reimplementing some of them using delegation. 

Would be a lot better if it was a Property has-a Hashtable (which later could be changed to another data structure), rather than a Property is-a Hashtable. Now to avoid breaking backward compatibility, it has remained this way for years :disappointed: 

So +1 if this can be fixed via composition :+1:

**Comment by kinow** [comment]
>Do we just have a naming convention issue here? E.g. when we have a CamelCase name, we expect it to be a class? If SuiteRuntimeClient is a factory method/function, should it have a lower_case name instead?

If it were a function, I think it would be less confusing, and users would not rely on the OO methods incorrectly. So that sounds like a plausible solution too :+1:

**Comment by oliver-sanders** [comment]
Closed by #3389
